rules/maprule.bzl

```diff
@@ -0,0 +1,536 @@
+# Copyright 2018 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Maprule implementation in Starlark.
+
+Maprule runs a specific command for each of the "foreach" source files. This allows processing
+source files in parallel, to produce some outputs for each of them.
+
+The name "maprule" indicates that this rule can be used to map source files to output files, and is
+also a reference to native.genrule that inspired this rule's design (though there are significant
+differences).
+
+This module exports:
+
+- The cmd_maprule() and bash_maprule() build rules.
+  They are the same except for the interpreter they use (cmd.exe and Bash respectively) and thus
+  the expected language of their "cmd" attribute. We will refer to them collectively as "maprule".
+
+- The maprule_testing struct. This should only be used by maprule's own unittests.
+
+Below you will find an Example, Tips and Tricks, and an FAQ.
+
+See the _ATTRS variable in this file for attribute documentation.
+
+## Example
```

Comment:
Drive by: Should the example (and some of this big string, go in the `doc` on the `rule()` itself so skydoc will find it and include it when generating docs?

Reply:
Actually, all I meant was fix how you documented things, I believe @c-parsons was working on getting stardoc going for things like this and avoiding the need for the extra copies.  So I'd suggest avoiding the duplication and just tweak your docs for now and let startdoc get wired up in an separate PR.

Reply:
I was more the wiring in of startdoc and cloning of `dicts.add` and `paths.dirname` you mentioned.  Since Chris had seemed to indicate in a past conversation the dep cycle would be fine, it just had to be carefully rolled out, I'd was thinking avoid letting him cover that since he seemed to have a plan for it.

Reply:
Sorry, I'm not sure I understood your comment -- do you suggest undoing the Stardoc wiring and copying of dicts/paths into maprule, or does @c-parsons have a plan so for now I can keep depending on dics.bzl/paths.bzl?

Reply:
> do you suggest undoing the Stardoc wiring and copying of dicts/paths into maprule

yes.

> does @c-parsons have a plan so for now I can keep depending on dics.bzl/paths.bzl

yes, that was my understanding.

My original comment was just to move all your docs/example text off the module doc and onto the actual apis, so when stardoc does get wired (by Chris), the docs for the two apis will be more complete.